radix-engine/src/engine/kernel.rs

```diff
@@ -717,6 +717,28 @@ where
 
         Ok(output)
     }
+
+    pub fn finalize(mut self, result: InvokeResult) -> TrackReceipt {
+        let final_result = match result {
+            Ok(res) => self.finalize_modules().map(|_| res),
+            Err(err) => {
+                // If there was an error, we still try to finalize the modules,
+                // but forward the original error (even if module finalizer also errors).
+                let _silently_ignored = self.finalize_modules();
+                Err(err)
+            }
+        };
```

Comment:
This section of code is a bit hard to read. Suggest:
```rust
let final_result = self.finalize_modules().and_then(|_| result);
self.track.finalize(final_result)
```

Reply:
I've reverted the change you suggested. It doesn't really work in the same way:
if both `result` and `finalize_modules` are `Err`, we want to return the earlier one, the one from `result` (and ignore the latter). While your snippet returns the latter Err (the result of `finalize_modules`) and ignores the one from `result`.

Reply:
Sounds good!